### PR TITLE
Show division in replays when available

### PR DESCRIPTION
**Pull request by BlackYps**
WIP
grafik

### CODE REVIEW

**Comment by BlackYps** [comment]
The core functionality is there, but it is still a bit rough around the edges.
There is a slight delay because we have to fetch the division info from the api, so at the moment it behaves a bit ugly visually.
I appreciate input how this could be solved better.
~~Similarly  the team title still displays (0) because I didn't manage to properly set the teamrating to null to trigger the removal of the rating display alltogether.~~
Edit: Fixed, but I don't know if the way I did it is good style

File: src/main/java/com/faforever/client/domain/GamePlayerStatsBean.java [hunk 1]
```diff
@@ -25,6 +25,7 @@ public class GamePlayerStatsBean {
   IntegerProperty team = new SimpleIntegerProperty();
   ObjectProperty<Faction> faction = new SimpleObjectProperty<>();
   ObjectProperty<OffsetDateTime> scoreTime = new SimpleObjectProperty<>();
+  ObjectProperty<GameOutcome> result = new SimpleObjectProperty<>();
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/domain/GamePlayerStatsBean.java]: I would probably just name this outcome

File: src/main/java/com/faforever/client/game/PlayerCardController.java [hunk 2]
```diff
@@ -98,10 +105,13 @@ public void initialize() {
...
         .bind(player.flatMap(PlayerBean::avatarProperty).map(avatarService::loadAvatar).when(showing));
+    divisionImageView.imageProperty()
+            .bind(division.flatMap(SubdivisionBean::smallImageUrlProperty).map(leaderboardService::loadDivisionImage).when(showing));
     playerInfo.textProperty()
         .bind(player.flatMap(PlayerBean::usernameProperty)
-            .flatMap(username -> rating.map(value -> i18n.get("userInfo.tooltipFormat.withRating", username, value))
-                .orElse(i18n.get("userInfo.tooltipFormat.noRating", username)))
+            .map(username -> (rating.get() != null && division.get() == null)
+                ? i18n.get("userInfo.tooltipFormat.withRating", username, rating.get())
+                : i18n.get("userInfo.tooltipFormat.noRating", username))
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/PlayerCardController.java]: This is not equivalent because before it was mapped off of rating so it would refresh based on changes to rating where as now we lost that property.

Also if we don't use the division why is it included in the if statement?

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 3]
```diff
@@ -43,6 +47,13 @@
 @Slf4j
 @RequiredArgsConstructor
 public class TeamCardController implements Controller<Node> {
+
+  private static final PseudoClass VICTORY = PseudoClass.getPseudoClass("victory");
+  private static final PseudoClass DEFEAT = PseudoClass.getPseudoClass("defeat");
+  private static final PseudoClass DRAW = PseudoClass.getPseudoClass("draw");
+  private static final PseudoClass UNKNOWN = PseudoClass.getPseudoClass("unknown");
+  private static final int UNSET = -1000000;
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]: No magic numbers please just use null

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 4]
```diff
@@ -51,19 +62,22 @@ public class TeamCardController implements Controller<Node> {
...
   private final ObjectProperty<RatingPrecision> ratingPrecision = new SimpleObjectProperty<>();
+  private final ObjectProperty<GameOutcome> teamResult = new SimpleObjectProperty<>();
   private final IntegerProperty teamId = new SimpleIntegerProperty();
   private final SimpleChangeListener<List<PlayerBean>> playersListener = this::populateTeamContainer;
   private final ObservableValue<Integer> teamRating = ratingProvider.flatMap(provider -> ratingPrecision.flatMap(precision -> players.map(playerBeans -> playerBeans.stream()
       .map(provider)
       .filter(Objects::nonNull)
       .map(rating -> precision == RatingPrecision.ROUNDED ? RatingUtil.getRoundedRating(rating) : rating)
-      .reduce(0, Integer::sum))));
+      .reduce(Integer::sum).orElse(UNSET))));
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]: null

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 5]
```diff
@@ -74,13 +88,18 @@ public void initialize() {
           case GameBean.OBSERVERS_TEAM -> i18n.get("game.tooltip.observers");
           default -> {
             try {
-              yield i18n.get("game.tooltip.teamTitle", id.intValue() - 1, teamRating);
+              if (teamRating == UNSET) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]: null

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 6]
```diff
@@ -107,6 +128,27 @@ private List<PlayerCardController> createPlayerCardControllers(List<PlayerBean>
       return controller;
     }).toList();
   }
+  
+  public void showGameResult() {
+    switch (teamResult.get()) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]: You should make sure the other pseudoClassStateChanges are set to false

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 7]
```diff
@@ -51,19 +62,22 @@ public class TeamCardController implements Controller<Node> {
...
   public Label teamNameLabel;
+  public Label gameResultLabel;
 
   private final ObjectProperty<List<Integer>> playerIds = new SimpleObjectProperty<>(List.of());
   private final ObjectProperty<List<PlayerBean>> players = new SimpleObjectProperty<>(List.of());
   private final ObjectProperty<Function<PlayerBean, Integer>> ratingProvider = new SimpleObjectProperty<>();
+  private final ObjectProperty<Function<PlayerBean, SubdivisionBean>> divisionProvider = new SimpleObjectProperty<>();
   private final ObjectProperty<Function<PlayerBean, Faction>> factionProvider = new SimpleObjectProperty<>();
   private final ObjectProperty<RatingPrecision> ratingPrecision = new SimpleObjectProperty<>();
+  private final ObjectProperty<GameOutcome> teamResult = new SimpleObjectProperty<>();
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]: Would name it teamOutcome

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 8]
```diff
@@ -300,6 +305,15 @@ private void onReplayChanged(ReplayBean newValue) {
     if (newValue.getReplayFile() != null) {
       enrichReplayLater(newValue.getReplayFile(), newValue);
     }
+    
+    newValue.setLeagueScores(null);
+    replayService.getLeagueScoreJournalForReplay(newValue)
+        .thenAccept(scores -> Platform.runLater(() -> {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Use the FxApplicationExecutor by passing it to the thenAcceptAsync method

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 9]
```diff
@@ -424,6 +438,10 @@ private CompletableFuture<Void> enrichReplayLater(Path path, ReplayBean replay)
   private void populateTeamsContainer(Map<String, List<GamePlayerStatsBean>> newValue) {
     CompletableFuture.supplyAsync(() -> createTeamCardControllers(newValue)).thenAcceptAsync(controllers -> {
       teamCardControllers.clear();
+      if (controllers.stream()
+          .map(TeamCardController::getTeamResult).noneMatch(gameOutcome -> gameOutcome != GameOutcome.DEFEAT)) {
+        controllers.forEach(teamCardController -> teamCardController.setTeamResult(GameOutcome.DRAW));
+      }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Why is this happening here instead of in createTeamCardControllers?

It seems like it should be done when the controllers are created rather than setting it to one value just to change it right after. I imagine we have all the information when the controllers are created.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 10]
```diff
@@ -300,6 +305,15 @@ private void onReplayChanged(ReplayBean newValue) {
...
       enrichReplayLater(newValue.getReplayFile(), newValue);
     }
+    
+    newValue.setLeagueScores(null);
+    replayService.getLeagueScoreJournalForReplay(newValue)
+        .thenAccept(scores -> Platform.runLater(() -> {
+          newValue.setLeagueScores(scores);
+          // This looks a bit ugly. Ideally we should wait with drawing the window until we have the league scores,
+          // then we don't need to trigger a redraw here
+          populateTeamsContainer(teams.getValue());
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: As far as I can tell the league scores are not even used in the populateTeamsContainer method so why do we need to call it again here?

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 11]
```diff
@@ -449,20 +469,53 @@ private List<TeamCardController> createTeamCardControllers(Map<String, List<Game
     }).toList();
   }
 
+  private GameOutcome calculateGameResult(Map<PlayerBean, GamePlayerStatsBean> statsByPlayer) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Just pass in a collection of GamePlayerStatsBean because we never use the keys

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 12]
```diff
@@ -449,20 +469,53 @@ private List<TeamCardController> createTeamCardControllers(Map<String, List<Game
...
+        .map(gameOutcome -> (gameOutcome == GameOutcome.MUTUAL_DRAW) ? GameOutcome.DRAW : gameOutcome)
+        .collect(Collectors.groupingBy(gameOutcome -> gameOutcome, Collectors.counting()));
+
+    if (outcomeCounts.containsKey(GameOutcome.VICTORY)) {
+      return GameOutcome.VICTORY;
+    }
+
+    return outcomeCounts.entrySet()
+        .stream()
+        .max(Entry.comparingByValue()).map(Entry::getKey).orElse(GameOutcome.UNKNOWN);
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Why is victory a special case? Shouldn't it be covered by the comparison below. Or it seems like there should be more special cases.

At the moment it isn't exactly clear why the outcome is being decided this way.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 13]
```diff
@@ -449,20 +469,53 @@ private List<TeamCardController> createTeamCardControllers(Map<String, List<Game
...
     GamePlayerStatsBean playerStats = statsByPlayerId.get(player);
     return playerStats == null ? null : playerStats.getFaction();
   }
 
   private Integer getPlayerRating(PlayerBean player, Map<PlayerBean, GamePlayerStatsBean> statsByPlayerId) {
     GamePlayerStatsBean playerStats = statsByPlayerId.get(player);
-    return playerStats == null ? null : playerStats.getLeaderboardRatingJournals()
+    if ((!replay.get().getLeagueScores().isEmpty()) || playerStats == null) {
+      return null;
+    }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Why do the league scores need to be empty for a player rating to be returned?

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java [hunk 14]
```diff
@@ -449,20 +469,53 @@ private List<TeamCardController> createTeamCardControllers(Map<String, List<Game
...
         .stream()
+        .filter(leagueScoreJournalBean -> leagueScoreJournalBean.getLoginId() == player.getId())
         .findFirst()
-        .filter(ratingJournal -> ratingJournal.getMeanBefore() != null)
-        .filter(ratingJournal -> ratingJournal.getDeviationBefore() != null)
-        .map(RatingUtil::getRating)
-        .orElse(null);
+        .map(LeagueScoreJournalBean::getDivisionBefore)
+        .orElse(null)
+    ).getValue();
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/replay/ReplayDetailController.java]: We shouldn't be using the replayProperty here because this just creates additional observables that aren't needed. We probably just want to pass in a map of leagueScoresByPlayer like we do for the rating.

File: src/test/java/com/faforever/client/builders/LeagueScoresListBuilder.java
Comment (Sheikah45) [inline src/test/java/com/faforever/client/builders/LeagueScoresListBuilder.java:1]: I don't think a list builder is really needed since all the methods here are exposed by the list class directly

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/domain/GamePlayerStatsBean.java (hunk 1 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/domain/GamePlayerStatsBean.java]: I named it this way, because that is how it is called in the api

Comment (Sheikah45) [reply src/main/java/com/faforever/client/domain/GamePlayerStatsBean.java]: I would still just name it outcome since that will make it clearer in this code what it represents.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 12 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Sometimes not all players report victory in the team that won. I think it happens when a player dies before the end of the match, then they report "defeat" for themselves. From my empirical testing I concluded that as soon as we detect a victory, we can assume that that team has won, even if everyone else in the team reports defeat

Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Wouldn't that be the case as well for draw? So does it make more sense rather than taking the mode taking the max reported result where Max is defined as priority?

File: src/main/java/com/faforever/client/game/PlayerCardController.java (hunk 2 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: when would the rating change? This is the rating associated with a particular replay, not the "live" rating.
 Losing the refresh was not a conscious decision though. These chained mappings elude me a little bit.
What do you mean we don't use the division? The intention is to hide the rating when there is a division instead.

File: src/main/java/com/faforever/client/game/TeamCardController.java (hunk 4 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/game/TeamCardController.java]: I can't use null here. If I do that the entire title disappears. Presumably because of a NPE when trying to map the teamrating.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 10 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: They are used because we create the teamCardControllers again. Then we set the divison provider in line 462. If I don't do this the division info is not displayed, because the additionial api call needs a moment to complete.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 13 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: I want to hide the player rating when we have league scores instead. The idea is that the league info is the new standard and rating is only used as a fallback

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 12 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: I don't really understand what you mean. Can you rephrase that?

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 9 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: I need to check if both controllers are reporting defeat. So I can't decide this in a single controller. Or do you mean to just move these lines as is to the end of the createTeamCardControllers function?

File: src/main/java/com/faforever/client/game/TeamCardController.java (hunk 6 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/game/TeamCardController.java]: I thought this was implicit because only one could be active. I can add setting the others to false

File: src/main/java/com/faforever/client/game/PlayerCardController.java (hunk 2 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: The rating would change if we changed the player value of the card.

And if the intention is to hide the rating when we have the division then we should modify the visible property in that case and not the actual value.

File: src/main/java/com/faforever/client/game/TeamCardController.java (hunk 4 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/TeamCardController.java]: Then it would still be better to handle that NPE where it occurs

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 14 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: So I change the start to this? `replay.get().getLeagueScores()` what exactly is the difference?

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 12 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: See if anyone reported victory, draw, defeat, unknown and take the first one because the default is everyone reports defeat so we only need to check if someone reported something other than that

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 13 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: As above then we should be modifying the visibility of the elements and not the values

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 14 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Using the property creates a bunch of intermediate objects that we never end up using.

File: src/main/java/com/faforever/client/game/PlayerCardController.java (hunk 2 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: And at a high level the maps are kind of like streams where map is used for converting a value while flat map is used for converting an observable value.

They create derivative observables that are efficient to subscribe to and react to changes upstream

Comment (BlackYps) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: but there is no visibility property to modify. The formating is done through using a different i18n key

Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: Then I would probably change it to a rating label and a division label to make it clearer rather than trying to use the same thing for both

File: src/main/java/com/faforever/client/game/TeamCardController.java (hunk 6 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/TeamCardController.java]: PsuedoClassStates are a list so you can have multiple active at the same time.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 9 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Is draw not separately reported from defeat by the API?
And if that is the case then we shouldn't set the team result in the card controller because it is misleading

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 10 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: We shouldn't have to recreate the teamCardControllers again. It should be handled like the showRatingChange where setting a value for the teamCardControllers will trigger the display rather than it needing to be recreated.

File: src/main/java/com/faforever/client/game/TeamCardController.java (hunk 4 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/TeamCardController.java]: You would need to add an OrElse mapping after the flatmap where it is used

File: src/main/java/com/faforever/client/game/PlayerCardController.java (hunk 2 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/game/PlayerCardController.java]: I would also generally say that we shouldn't be returning null from methods like getPlayerRating just because the division is set. The display of the values should be happening at a higher level

**Comment by Sheikah45** [comment]
@BlackYps were you ever going to get back to this?

There have also been a few changes to the underlying objects as an FYI.

**Comment by BlackYps** [comment]
I don't plan to abandon this, on the other hand I am really not motivated to do the refactoring that is required here.
So I guess the honest answer is, I really don't know...

**Comment by Sheikah45** [comment]
Then I might just close this in a week and it can be reopened when you revisit it.

**Comment by BlackYps** [comment]
I'm picking this up again

File: src/main/java/com/faforever/client/mapstruct/ReplayMapper.java [hunk 15]
```diff
@@ -136,8 +145,8 @@ default HashMap<String, List<GamePlayerStats>> mapTeamStatsFromParser(ReplayData
                                                                               null);
         Float factionFloat = (Float) armyInfo.get("Faction");
         Faction faction = Faction.fromFaValue(factionFloat.intValue());
-        GamePlayerStats stats = new GamePlayerStats(player, (byte) 0, team, faction, null,
-                                                    List.of(ratingJournal));
+        GamePlayerStats stats = new GamePlayerStats(false, faction, (byte) 0, team, (byte) 0, (byte) 0, null,
```
Comment (BlackYps) [inline src/main/java/com/faforever/client/mapstruct/ReplayMapper.java:148]: What is the purpose of the `mapTeamStatsFromParser` function? I noticed that the GamePlayerStats only used some of the fields that were provided by the api, so I expanded it, but this function creates a GamePlayerStats object and feeds it data from a parsed replay, but I don't know which of the additional fields can be extracted from the parser.

**Comment by BlackYps** [comment]
The population of the values of the controllers and the displaying of these values are now nicely decoupled. There is one issue though. In a matchmaker game the people that don't have a division still display the rating and the team rating is displayed as well. The problem is that the player cards without a division don't know about the presence of divisions in different player cards. We need some way to check if we found any league stats for the replay (the api call) and then disable rating display for the whole thing somehow

File: src/main/java/com/faforever/client/mapstruct/ReplayMapper.java (hunk 15 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/mapstruct/ReplayMapper.java:148]: It is exactly as you say that it populates the team stats for local replay files. Local replays don't have all the data from the api so only fill in some fields

**Comment by Sheikah45** [comment]
For displaying the whole thing you can just have a display type enum that has a property in the player card. Then you can set that property appropriately to display either rating or league. Could even just bind it to some property in the parent.

File: src/main/java/com/faforever/client/game/TeamCardController.java [hunk 16]
```diff
@@ -70,18 +85,21 @@ public class TeamCardController extends NodeController<Node> {
...
             } catch (NumberFormatException e) {
               yield "";
             }
           }
-        })));
-
+        }));
+    JavaFxUtil.bindManagedToVisible(teamRatingLabel, gameResultLabel);
+    teamRatingLabel.textProperty().bind(teamRating.map(value -> i18n.get("game.tooltip.ratingFormat", value)));
+    teamRatingLabel.visibleProperty().bind(Bindings.createBooleanBinding(() -> teamRating.getValue() != null && teamRating.getValue() != 0, teamRating));
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/TeamCardController.java]:
```suggestion
    teamRatingLabel.visibleProperty().bind(teamRating.map(value -> value != 0));
```

File: src/main/java/com/faforever/client/mapstruct/ReplayMapper.java (hunk 15 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/mapstruct/ReplayMapper.java:148]: so how should I deal with this situation? Put in some dummy data like I did here?

**Comment by BlackYps** [comment]
I guess a boolean would even be enough, something like `gameHasLeagueScore`. I don't have to differentiate multiple states, just this binary decision

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 12 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Ah yeah, I see now what you mean. I think you are correct. So that means I don't even need to count, but I just need to check the presence of results in that order.

File: src/main/java/com/faforever/client/replay/ReplayDetailController.java (hunk 9 above)
Comment (BlackYps) [reply src/main/java/com/faforever/client/replay/ReplayDetailController.java]: Months ago when I was starting to develop this, I found a game that reported me that both teams were defeated. I didn't save the replay Id, so I can't check now. It might have been because of my own faulty calculations at that time. E.g. only a minority reported a draw, so the majority of defeat results got chosen instead.
There is no way to know now, so I will remove this check and if we ever run into a replay that reports both teams as defeated we can investigate again.

**Comment by Sheikah45** [comment]
Yeah I was just thinking of an enum to make it clearer what it signifies. Since hasLeagueScore doesn't imply as much that ratings won't be shown.

Where as something like display type would make that clearer.

**Comment by BlackYps** [comment]
yeah, I can see that.

**Comment by Sheikah45** [comment]
Was testing this out and one thing that stuck out as a little bit odd to me. Why do we require pressing the show game result to show the game outcome on the team card. Maybe we should just display it always on load?

File: src/main/java/com/faforever/client/mapstruct/ReplayMapper.java (hunk 15 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/mapstruct/ReplayMapper.java:148]: Yeah that is fine

**Comment by BlackYps** [comment]
I think this kind as introduced to make it possible to watch a game without being spoiled on which side wins

**Comment by Sheikah45** [comment]
Sure but now you can also play directly from the search result page

**Comment by BlackYps** [comment]
Hmm, that's true. Maybe we should discuss this with more users to better understand how they use this button at the moment. It's definitely an interesting suggestion, but imo it is unrelated to this PR

**Comment by Sheikah45** [comment]
What is the status of this?

**Comment by BlackYps** [comment]
From my pov it is ready

**Review by Sheikah45 — COMMENTED** [review]
The only consequential thing I see is a race condition between populateTeamContainers and loading the league score results.

**Comment by BlackYps** [comment]
Can I do something about that race condition?

**Comment by Sheikah45** [comment]
You could make sure that you don't try to load the league data until after the team containers are populated

**Comment by BlackYps** [comment]
I think I first need to understand the race condition a little better. As I see it we run the risk that the league info is applied to the old teamcardcontrollers, so they get cleared right after and new controllers for the new replay get added in populateTeamsContainer if this happens after the application of the league info.
Is this the correct one, or did you have a different race condition in mind?
And why is createTeamCardControllers called asyncronously? I do not immediately see any operation in that method that would make this necessary.

**Comment by Sheikah45** [comment]
They are created asynchronously so they don't block the application thread and cause the ui to freeze. And to cure the race condition you would just load the league info after create team controllers is called or make it apart of create team controllers

**Review by Sheikah45 — APPROVED** [review]
There was an issue with the team card controllers being added twice but I fixed it.

**Comment by RoraRaven** [comment]
Showing division is nice, but was it necessary to remove the rating change? It means you need to check your profile after every match and keep a spreadsheet if you want to track changes rather than it being plainly displayed in the replays.

**Comment by BlackYps** [comment]
Yes, the divisions are supposed to replace the rating. The rating system is considered to be part of the backend, it's not necessary to show it to the user when the division system exists

**Comment by RoraRaven** [comment]
Are we going to get transparency in how divisions are calculated? Currently divisions are more or less meaningless, every other system (ingame player lists, lobby entry requirements, matchmaking, game quality rating) works on rating.

Divisions are really large and imprecise, you could win 15 games in a row and see no change. Even systems that obfuscate Elo / MMR (eg LoL) show points to measure how far up or down you are inside a division.

**Comment by BlackYps** [comment]
What exactly is the problem you are having? I'm trying to und erstand it so I can hopefully address it in my answer.
And did you have a look at the leaderboard tab? Your comment gives me the impression that you didn't, because you can see points there and it's not possible to win 15 games in a row without changing divisions, because each division is only 10 points big.

**Comment by RoraRaven** [comment]
I wasn't aware about the Leaderboards or Score system at all. I was under the impression that division was based on rating (e.g. Gold V being 800 - 900).

This does answer my questions.

I still don't understand why you would hide information from the player, even if you consider rating to be back-end, but that's your prerogative.
